### PR TITLE
List devices

### PR DESCRIPTION
**Pull request by rishubhjain**
@aravindavk Please review

### CODE REVIEW

File: plugins/device/rest.go [hunk 1]
```diff
@@ -104,3 +104,28 @@ func deviceAddHandler(w http.ResponseWriter, r *http.Request) {
...
+	logger := gdctx.GetReqLogger(ctx)
+	peerID := mux.Vars(r)["peerid"]
+	if uuid.Parse(peerID) == nil {
+		restutils.SendHTTPError(ctx, w, http.StatusBadRequest, "Invalid peer-id passed in url")
+		return
+	}
+
+	devices, err := GetDevices(peerID)
+	if err != nil {
+		logger.WithError(err).WithField("peerid", peerID).Error(err)
```
Comment (prashanthpai) [inline plugins/device/rest.go]: Pass an error string to Error() such as - `Failed to get devices for peer`

File: plugins/device/rest.go [hunk 2]
```diff
@@ -104,3 +104,28 @@ func deviceAddHandler(w http.ResponseWriter, r *http.Request) {
...
+
+	ctx := r.Context()
+	logger := gdctx.GetReqLogger(ctx)
+	peerID := mux.Vars(r)["peerid"]
+	if uuid.Parse(peerID) == nil {
+		restutils.SendHTTPError(ctx, w, http.StatusBadRequest, "Invalid peer-id passed in url")
+		return
+	}
+
+	devices, err := GetDevices(peerID)
```
Comment (prashanthpai) [inline plugins/device/rest.go]: Unrelated to this PR: I had a look at `GetDevices` which does `return nil, nil` at the end. That can be improved as follows:

```go
func GetDevices(peerID string) ([]deviceapi.Info, error) {

        peerInfo, err := peer.GetPeer(peerID)
        if err != nil {
                return nil, err
        }

        var deviceInfo []deviceapi.Info
        if devices, ok := peerInfo.MetaData["devices"]; ok {
                if err := json.Unmarshal([]byte(devices), &deviceInfo); err != nil {
                        return nil, err
                }
        }

        return deviceInfo, nil
}
```

As you have to update the PR to address the comment below, you might as well fix this one as well.

**Comment by prashanthpai** [comment]
The REST API documentation needs to be regenerated.

File: plugins/device/init.go [hunk 3]
```diff
@@ -27,6 +27,13 @@ func (p *Plugin) RestRoutes() route.Routes {
 			RequestType:  utils.GetTypeString((*deviceapi.AddDeviceReq)(nil)),
 			ResponseType: utils.GetTypeString((*deviceapi.AddDeviceResp)(nil)),
 			HandlerFunc:  deviceAddHandler},
+		route.Route{
+			Name:         "DeviceList",
+			Method:       "GET",
+			Pattern:      "/peers/{peerid}/devices",
```
Comment (aravindavk) [inline plugins/device/init.go]: can this be changed to `/devices/{peerid}` and `/devices`, when peerid not specified then return devices from all peers else only from that peer.

For uniformity, we can change `POST /peers/{peerid}/devices` to `POST /devices/{peerid}`

File: plugins/device/rest.go (hunk 2 above)
Comment (prashanthpai) [reply plugins/device/rest.go]: @rishubhjain The code on `GetDevices` uses `devices` as key, it's `_devices` at other places. Please find places where it's `devices` and change them to `_devices`

**Comment by rishubhjain** [comment]
@prashanthpai If possible could you check again, I have used _devices only.

File: plugins/device/init.go (hunk 3 above)
Comment (rishubhjain) [reply plugins/device/init.go]: Will address listing of all devices in another PR as it requires some discussion

File: plugins/device/init.go [hunk 4]
```diff
@@ -22,11 +22,18 @@ func (p *Plugin) RestRoutes() route.Routes {
 		route.Route{
 			Name:         "DeviceAdd",
 			Method:       "POST",
-			Pattern:      "/peers/{peerid}/devices",
+			Pattern:      "/devices/{peerid}",
```
Comment (aravindavk) [inline plugins/device/init.go]: `restclient` patch merged recently. please update this change in `$SRC/pkg/restclient/device.go`

File: plugins/device/rest.go [hunk 5]
```diff
@@ -106,3 +106,52 @@ func deviceAddHandler(w http.ResponseWriter, r *http.Request) {
...
+			restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+		}
+		return
+	}
+
+	restutils.SendHTTPResponse(ctx, w, http.StatusOK, devices)
+
+}
+
+func listAllHandler(w http.ResponseWriter, r *http.Request) {
```
Comment (aravindavk) [inline plugins/device/rest.go]: you can remove this function if planning to implement this later.

File: plugins/device/rest.go [hunk 6]
```diff
@@ -106,3 +106,52 @@ func deviceAddHandler(w http.ResponseWriter, r *http.Request) {
...
+
+	ctx := r.Context()
+	logger := gdctx.GetReqLogger(ctx)
+	peers, err := GetPeers()
+	if err != nil {
+		logger.WithError(err).Error("Failed to get peers from store")
+	}
+
+	devices, err := GetDevices(peerID)
```
Comment (aravindavk) [inline plugins/device/rest.go]: loop is missing. peerID is undefined

**Pull request merged**